### ISSUE TITLE
Deleted C++ object: WebEngineView

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#3498]
Seeing crash reports for some `RuntimeError` exceptions when a tab was closed before it finished loading in Qt 5.10, probably somehow related to #3427...

I can see three exceptions:

# When getting title

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/browsertab.py", line 761, in _on_load_finished
    if not self.title():
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/webengine/webenginetab.py", line 615, in title
    return self._widget.title()
RuntimeError: wrapped C/C++ object of type WebEngineView has been deleted
```

https://crashes.qutebrowser.org/view/c8cbd1bc
https://crashes.qutebrowser.org/view/1cfd8512
https://crashes.qutebrowser.org/view/f0367193
https://crashes.qutebrowser.org/view/f5c1f5cb
https://crashes.qutebrowser.org/view/06b83479
https://crashes.qutebrowser.org/view/7b9ecafa

# When getting page (to get the URL)

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/qutebrowser/browser/browsertab.py", line 756, in _on_load_finished
    if self.url().scheme() == 'https':
  File "/usr/local/lib/python3.6/site-packages/qutebrowser/browser/webengine/webenginetab.py", line 611, in url
    page = self._widget.page()
RuntimeError: wrapped C/C++ object of type WebEngineView has been deleted
```

https://crashes.qutebrowser.org/view/59914c37
https://crashes.qutebrowser.org/view/f6f3f0fc
https://crashes.qutebrowser.org/view/c288b1c3
https://crashes.qutebrowser.org/view/010849dd

# When setting zoom factor

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/webengine/webenginetab.py", line 568, in _restore_zoom
    self.zoom.set_factor(self._saved_zoom)
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/browsertab.py", line 309, in set_factor
    self._set_factor_internal(factor)
  File "/usr/lib/python3.6/site-packages/qutebrowser/browser/webengine/webenginetab.py", line 446, in _set_factor_internal
    self._widget.setZoomFactor(factor)
RuntimeError: wrapped C/C++ object of type WebEngineView has been deleted
```

https://crashes.qutebrowser.org/view/a3193ff3
https://crashes.qutebrowser.org/view/fd0d3a35
https://crashes.qutebrowser.org/view/b568a04a

Interestingly, this has happened to someone before Qt 5.10 too:

https://crashes.qutebrowser.org/view/6787deba
https://crashes.qutebrowser.org/view/d90bd26b

**Comment by The-Compiler** [comment]
@qutebrowser-bot huh. 😆

**Comment by The-Compiler** [comment]
Sorry for the noise @Penaz91! Not sure how that happened 😆 I turned that feature off now.

As for this issue: Not sure whether #3427 is really responsible, I *think* there are crash reports there which are from before that (but with Qt 5.10).

**Comment by Penaz91** [comment]
No worries, it probably called me because I have a Qt/C++ respository.
I have no idea if said pull is responsible, i'm still running Qt 5.9.3 and everything's stable

**Comment by The-Compiler** [comment]
Hopefully fixed for v1.1.1 and v1.2.0. I'll probably have to cut a release soon-ish...

**Comment by toofar** [comment]
I was having a ponder about the general case of handling scheduled calls on deleted tabs while looking at https://github.com/qutebrowser/qutebrowser/issues/3895 where I applied the same `if deleted: return` guard as applied in several places already.

The approach to dealing with these errors has been "do nothing", or "just return". I mostly work on web backends at the day job, and when we run into a case like that we raise an exception that is handled by some global exception handler which just returns a 500 or something and proceeds with handling the next request. Could we do something like that here?

The three stack traces in the top post start from the loadFinished signal handler, the one in #3895 comes from a SingleShot timer, and loadFinished before that. Same with https://github.com/qutebrowser/qutebrowser/pull/4019. https://github.com/qutebrowser/qutebrowser/commit/b130162d56b is from a SingleShot timer from a search. https://github.com/qutebrowser/qutebrowser/commit/ee2c7658592d65 was apparently added pre-emptively.

So unfortunately it doesn't look like we can just put the guard in one single signal handler and before fanning out, there are a couple of ways these calls can be deferred. So what do we have in the way of a global event loop where we could potentially handle this case after it happened, instead of guarding for it in many places?

The stack trace when one of these exceptions happens and isn't handled looks like this:

```
Current thread 0x00007f501e16f100 (most recent call first):
  File "/home/user/projects/qutebrowser/qutebrowser/app.py", line 126 in qt_mainloop
  File "/home/user/projects/qutebrowser/qutebrowser/app.py", line 116 in run
  File "/home/user/projects/qutebrowser/qutebrowser/qutebrowser.py", line 234 in main
  File "/home/user/projects/qutebrowser/qutebrowser/__main__.py", line 15 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main
```

Okay, so we are in the Qt event loop. Is there a global exception handler we can hook into there? I don't image it works at all with python exceptions, a web search certainly didn't come up with anything. Does PyQt have any magic that can help in this case? I don't know, that would probably be the way to go if so. But python itself does have a global exception handler. In fact the reason for the whole program crashing so ungraecfully with a segfault seems to because that exception handler and Qt/PyQt aren't integrated at all. So can we identify these exceptions in the python except hook and have our "do nothing" logic there? Actually, it looks like we are already ignoring some errors in [exactly that fashion](https://github.com/qutebrowser/qutebrowser/blob/0b221d3/qutebrowser/misc/crashsignal.py#L245).

```diff
diff --git i/qutebrowser/misc/crashsignal.py w/qutebrowser/misc/crashsignal.py
index 1b31536f84ba..9a2b522a455f 100644
--- i/qutebrowser/misc/crashsignal.py
+++ w/qutebrowser/misc/crashsignal.py
@@ -219,7 +219,7 @@ class CrashHandler(QObject):
             True: Exception hook should be aborted.
             False: Continue handling exception.
         """
-        exctype, _excvalue, tb = exc
+        exctype, excvalue, tb = exc

         if not self._quitter.quit_status['crash']:
             log.misc.error("ARGH, there was an exception while the crash "
@@ -246,6 +246,14 @@ class CrashHandler(QObject):
                            "Please report this as a bug.")
             return True

+        if exctype is RuntimeError and excvalue.args[0] == (
+            "wrapped C/C++ object of type WebEngineView has been deleted"
+        ):
+            # Assume a task was executed async and referred to a webview that had
+            # been deleted. This case hopefully isn't fatal for the whole browser
+            log.misc.error("Ignoring deleted WebEngineView exception and retuning to QT main loop.")
+            return True
+
         return False

     def exception_hook(self, exctype, excvalue, tb):
```

Well, that patch works fine with my reproducer from https://github.com/qutebrowser/qutebrowser/commit/c32f5afcc43092de49c0597a0bfd5a1ddf0363ae! The worry here is that we ignore exceptions which should be fatal. So we have to have some confidence that our heuristic will only catch things that are ignorable. Here I'm just looking at the error message, is that good enough? I don't know. The only other thing that I can think of that we can examine is the traceback, which has the frames with code location and locals. But if we start parsing that we are basically back to doing case-by-case guards.

In general what should our heuristic be checking for anyway? Is it safe to ignore any error relating to a specific tab? We have lots of tabs, one of them dying shouldn't bring down the application if possible. But what about something like an IO error that causes all your tabs to crash, should we be ignoring that? Maybe? Anyhow, I'm not sure what other cases we might want to give the same treatment. I searched for
